### DISCUSSION TITLE
[Help wanted] TextBox issue with device virtual keyboard

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#12382] [Q&A] [answered]
Hi, 
it's probably a question for @Gillibald

My goal is to create a RichTextBox based on TextBox and Topten Software's RichTextBox, but my question is more about the Avalonia TextBox. When testing on Android, I noticed that my textbox is not working correctly with Android's virtual device.

**So I did the minimal test, I just duplicated Avalonia's TextBox theme and class, and renamed to TextBoxEx.
I also had to copy a few internal dependencies (StringBuilderCache, StringUtils, UndoRedoHelper, and TextBoxTextInputMethodClient), and call two internal or private method with reflection. I also used System.Reactive instead of Avalonia own implementation**

So basically, I should have the exact same TextBox as Avalonia. However, when I try on Android, the behavior with the virtual keyboard is different. Avalonia's TextBox work fine, but the my copy of Avalonia's TextBox do not work well. For example, when I double-tab a word in the TextBox, and try to delete, it do weird stuffs (erase a single char, or moving the caret at another position).

Do you have any idea of what could be wrong?

Thank you for any help.

**Comment by Gillibald** [comment]
You need to handle this event: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBox.cs#L326

**Reply by dbriard** [reply]
@Gillibald Yes, it is done as I my debug test I copy/pasted the TextBox of Avalonia and just renamed it.
The virtual keyboard appear, but the behavior is not the same as TextBox while the code is exactly the same... Build in TextBox works fine, but copied class has bugs :(

May be there is some code somewhere that has hardcoded TextBox type?

**Reply by Gillibald** [reply, marked as answer]
You need to adjust the TextBoxTextInputMethod Client implementation. It only works with Avalonia's TextBox. You are most likely not using a TextPresenter but the default implementation is expecting it

**Reply by dbriard** [reply]
@Gillibald 
I restarted again, and found the issue, I forgot to replace TextBox at 2 places here: :(
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/d7a6759b-19b0-42ac-bb8d-cb9338a16f69)

Now it works fine, sorry to have disturbed you!
